login1/dbus.go

```diff
@@ -74,6 +74,141 @@ func (c *Conn) initConnection() error {
 	return nil
 }
 
+// Session object definition.
+type Session struct {
+	ID   string
+	UID  uint32
+	User string
+	Seat string
+	Path dbus.ObjectPath
+}
+
+// User object definition.
+type User struct {
+	UID  uint32
+	Name string
+	Path dbus.ObjectPath
+}
+
+func (s Session) toInterface() []interface{} {
+	return []interface{}{s.ID, s.UID, s.User, s.Seat, s.Path}
+}
+
+func sessionFromInterfaces(session []interface{}) (*Session, error) {
+	id, ok := session[0].(string)
+	if !ok {
+		return nil, fmt.Errorf("failed to typecast session field 0 to string")
+	}
+	uid, ok := session[1].(uint32)
+	if !ok {
+		return nil, fmt.Errorf("failed to typecast session field 1 to uint32")
+	}
+	user, ok := session[2].(string)
+	if !ok {
+		return nil, fmt.Errorf("failed to typecast session field 2 to string")
+	}
+	seat, ok := session[3].(string)
+	if !ok {
+		return nil, fmt.Errorf("failed to typecast session field 2 to string")
+	}
+	path, ok := session[4].(dbus.ObjectPath)
+	if !ok {
+		return nil, fmt.Errorf("failed to typecast session field 4 to ObjectPath")
+	}
+
+	ret := Session{ID: id, UID: uid, User: user, Seat: seat, Path: path}
+	return &ret, nil
+}
+
+func userFromInterfaces(user []interface{}) (*User, error) {
+	uid, ok := user[0].(uint32)
```

Comment:
Please check for `len(user) >= 3` before starting to index-access the array.

Reply:
Should we return an error if the number of fields don't match?

Reply:
Same as above, yes.